packages/backend/src/apps/paysg/actions/create-payment/index.ts

```diff
@@ -0,0 +1,169 @@
+
+type CreatePaymentPayload = {
+  reference_id: string
+  payer_name: string
+  payer_address: string
+  payer_identifier: string
+  payer_email: string
+  description: string
+  amount_in_cents: number
+  metadata: Record<string, string>
+  due_date?: string
+  return_url?: string
+}
+
+function constructPayload(parameters: IJSONObject): CreatePaymentPayload {
+  const payload: CreatePaymentPayload = {
+    reference_id: parameters.referenceId as string,
+    payer_name: parameters.payerName as string,
+    payer_address: parameters.payerAddress as string,
+    payer_identifier: parameters.payerIdentifier as string,
+    payer_email: parameters.payerEmail as string,
+    description: parameters.description as string,
+    amount_in_cents: Number(parameters.paymentAmountCents),
+    metadata: Object.create(null),
+  }
+
+  if (parameters.dueDate) {
+    payload['due_date'] = parameters.dueDate as string
+  }
+
+  if (parameters.returnUrl) {
+    payload['return_url'] = parameters.returnUrl as string
+  }
+
+  // FIXME (ogp-weeloong): by default, we populate metadata with 1 empty row
+  // even if its optional, for UX reasons. for now, account for this case in
+  // code until we make the necessary UX changes to not need that 1 empty row.
+  const metadata = ((parameters.metadata as IJSONObject[] | null) ?? []).filter(
+    (metadatum) => !!metadatum.key,
+  )
+  if (metadata?.length) {
+    for (const metadatum of metadata) {
+      const { key, value } = metadatum as { key: string; value: string }
+      payload['metadata'][key] = value
+    }
+  }
+
+  return payload
+}
+
+const action: IRawAction = {
+  name: 'Create Payment',
+  key: 'createPayment',
+  description: 'Create a new PaySG payment',
+  arguments: [
+    {
+      label: 'Reference ID',
+      key: 'referenceId',
+      type: 'string' as const,
+      required: true,
+      variables: true,
+    },
+    {
+      label: 'Payer Name',
+      key: 'payerName',
+      type: 'string' as const,
+      required: true,
+      variables: true,
+    },
+    {
+      label: 'Payer Address',
+      key: 'payerAddress',
+      type: 'string' as const,
+      required: true,
+      variables: true,
+    },
+    {
+      label: 'Payer Identifier',
+      description: 'e.g. NRIC',
+      key: 'payerIdentifier',
+      type: 'string' as const,
+      required: true,
+      variables: true,
+    },
+    {
+      label: 'Payer Email',
+      key: 'payerEmail',
+      type: 'string' as const,
+      required: true,
+      variables: true,
+    },
+    {
+      label: 'Description',
+      key: 'description',
+      type: 'string' as const,
+      required: true,
+      variables: true,
+    },
+    {
+      label: 'Payment amount (in cents)',
+      key: 'paymentAmountCents',
+      type: 'string' as const,
+      required: true,
+      variables: true,
+    },
+    {
+      label: 'Due Date',
+      description: 'Must be formatted as DD-MM-YYYY (e.g. 31-DEC-2023)',
```

Comment:
this due date not FormSG friendly. perhaps we stick to DD MMM YYYY (e.g. 21 Jan 2023)

Reply:
and the response date from paysg should be converted as such

Reply:
good call, done